server/src/main/java/org/elasticsearch/action/admin/indices/refresh/TransportShardRefreshAction.java

```diff
@@ -84,53 +75,37 @@ protected ReplicationResponse newResponseInstance(StreamInput in) throws IOExcep
     protected void shardOperationOnPrimary(
         BasicReplicationRequest shardRequest,
         IndexShard primary,
-        ActionListener<PrimaryResult<BasicReplicationRequest, ReplicationResponse>> listener
+        ActionListener<PrimaryResult<RefreshReplicationRequest, ReplicationResponse>> responseListener
```

Comment:
Now, I believe this is more elegant: we do not need to care about sending manually actions to unpromotable shards. You just specify whether to send the replicated actions to either promotable/unpromotable shards using the ReplicaForwardOptions and just code the logic in shardOperationOnReplica() below.